lib/scripting/grammar/grammar.ts

```diff
@@ -0,0 +1,268 @@
+/*
+ * VPDB - Virtual Pinball Database
+ * Copyright (C) 2019 freezy <freezy@vpdb.io>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+import { Grammars, IToken, Parser } from 'ebnf';
+import { generate } from 'escodegen';
+import { Program, Statement } from 'estree';
+import { getTextFile } from '../../refs.node';
+import { program } from '../estree';
+import { ppArray } from '../post-process/array';
+import { ppAssign } from '../post-process/assign';
+import { ppBranch } from '../post-process/branch';
+import { ppCall } from '../post-process/call';
+import { ppClass } from '../post-process/class';
+import { ppConditional } from '../post-process/conditional';
+import { ppConst } from '../post-process/const';
+import { ppError } from '../post-process/error';
+import { ppExpr } from '../post-process/expr';
+import { ppHelpers } from '../post-process/helpers';
+import { ppLiteral } from '../post-process/literal';
+import { ppLoop } from '../post-process/loop';
+import { ppMethod } from '../post-process/method';
+import { ppVarDecl } from '../post-process/vardecl';
+import { ppWith } from '../post-process/with';
+
+const dashAst = require('dash-ast');
+
+export interface ESIToken extends IToken {
+	estree: any;
+	parent: ESIToken;
+	children: ESIToken[];
+}
+
+export class Grammar {
+	private TOKEN_TERMINAL_KEYWORDS = 'Keywords ::= ';
+
+	private GRAMMAR_TARGET_FORMAT = 'Format';
+	private GRAMMAR_TARGET_PROGRAM = 'Program';
+
+	private parser: Parser;
+	private keywords: { [index: string]: string } = {};
+
+	private postProcessors = [
+		ppHelpers,
+		ppLiteral,
+		ppExpr,
+		ppVarDecl,
+		ppConst,
+		ppArray,
+		ppError,
+		ppAssign,
+		ppConditional,
+		ppLoop,
+		ppBranch,
+		ppWith,
+		ppMethod,
+		ppCall,
+		ppClass,
+	];
+
+	constructor() {
+		let grammar = getTextFile('grammar.bnf');
+
+		this.setKeywords(grammar);
+
+		grammar = this.addCaseInsensitiveKeywords(grammar);
```

Comment:
Currently we only patch 'Keywords'. I wanted the EBNF to be readable, and I wanted this to be the exact set of keywords with the case that matches the documentation.

`Keywords ::= 'And' | 'ByVal' | 'ByRef' | 'Case' | 'Call' | 'Class' | 'Const' | 'Default' | 'Dim' | 'Do' | 'Each' | 'ElseIf' | 'Else' | 'Empty' | 'End' | 'Erase' | 'Error' | 'Eqv' | 'Exit' | 'Explicit' | 'False' | 'For' | 'Function' | 'Get' | 'GoTo' | 'If' | 'In' | 'Is' | 'Let' | 'Loop' | 'Mod' | 'New' | 'Next' | 'Nothing' | 'Not' | 'Null' | 'On' | 'Option' | 'Or' | 'Preserve' | 'Private' | 'Property' | 'Public' | 'ReDim' | 'Resume' | 'Select' | 'Set' | 'Step' | 'Sub' | 'Then' | 'To' | 'True' | 'Until' | 'While' | 'WEnd' | 'With' | 'Xor'`

To allow the formatter to handle all the different cases, it get's rewritten as `(A|a)(N|n)(D|d) | (B|b)(Y|y)....`, etc.

Also, when it's patching the Keywords node, it's creating a lookup table.

So during the reformat,  if someone coded `byval`, it rewrites as `ByVal` because of the lookup table. 

The other grammar rules require exact case to match:

`ParameterModifier ::= 'ByVal' | 'ByRef'`

This way we don't have:

`ParameterModifier ::= '(B|b)(Y|y)(V|v)(A|a)(L|l)' 

This was all for readability and to be able to easily use the https://menduz.com/ebnf-highlighter/ tool.

Reply:
Alright, thanks for the explanation, makes sense.